src/ACMESharp/Crypto/JOSE/JWSAlgorithm.cs

```diff
@@ -0,0 +1,72 @@
+using System;
+
+namespace ACMESharp.Crypto.JOSE
+{
+    public class JWSAlgorithm
+    {
+        private IJwsSigner _jwsTool;
+
+        public JWSAlgorithm(string jwsAlgorithm)
+        {
+            if (!int.TryParse(jwsAlgorithm.Substring(2), out int size))
+                throw new ArgumentException("Could not parse Key or Hash size", nameof(jwsAlgorithm));
+
+            if (jwsAlgorithm.StartsWith("ES"))
+            {
+                var tool = new Impl.ESJwsSigner(size);
+
+                _jwsTool = tool;
+            }
+
+            if (jwsAlgorithm.StartsWith("RS"))
+            {
+                var tool = new Impl.RSJwsSigner(size);
+
+                _jwsTool = tool;
+            }
+
+            if (_jwsTool == null)
+                throw new InvalidOperationException("Unknown JwsAlgorithm");
+        }
+
+        public JWSAlgorithm(JwsAlgorithmExport exported)
+            :this(exported.Algorithm)
+        {
+            _jwsTool.Import(exported.PrivateKey);
+        }
+
+        public string JwsAlg => _jwsTool.JwsAlg;
+        
+        public void Dispose()
+        {
+            _jwsTool.Dispose();
+        }
+
+        public JwsAlgorithmExport Export()
+        {
+            var export = new JwsAlgorithmExport
+            {
+                Algorithm = _jwsTool.JwsAlg,
```

Comment:
I wonder if should derive the `Algorithm` value from the original constructor argument instead of what the IJwsSigner is giving you.  In this case the `JwsAlg` value is what is needed for a JWS payload to understand the contents and associated key type, but it may not exactly translate to the specific IJwsSigner implementation upon reload.

And along those same lines, the `JwsAlgorithm(string)` constructor would actually be taking an *implementation* identifier of sorts, not actually a JWS algorithm ID.

For example, in the current core code base, there are the 2 default JWS signer implementations, RSA and ECDSA, but I recently had to add a third one, an alternative ECDSA version because the  2 default implementations would not work in my target environment because they relied on the BCL crypto support which was missing, and I needed to add an alternative variation that was based on BouncyCastle.  Even though it's a distinct implementation of it's own, it still implements the ES256 (and ES384 and ES521) algorithms which is what it returns in its `JwsAlg` property.

Reply:
You are right - I reduced the extend-ability here, which should be there to swap out the actual JWS implementation. I was so fixated on making reuse of the key more "graspable" and straight forward, that I ignored that.
I have an idea, where you would get the possibility to define which JwsImplementation reacts on which name - something like a Dict<string, Func<string, IJwsSigner>> or something along those lines. Also instead of using the JwsAlg-Names i could use a more distinguished list like DefaultRSA374 or something like that.